python/ray/llm/_internal/batch/processor/base.py

```diff
@@ -0,0 +1,199 @@
+        """Execute the processor:
+        preprocess -> stages -> postprocess.
+        Note that the dataset won't be materialized during the execution.
+
+        Args:
+            dataset: The input dataset.
+
+        Returns:
+            The output dataset.
+        """
+        if self.preprocess is not None:
+            dataset = dataset.map(self.preprocess)
+
+        for stage in self.stages.values():
+            # Prepare .map_batches() arguments.
+            kwargs = stage.map_batches_kwargs.copy()
+            kwargs["batch_size"] = self.config.batch_size
+            kwargs.update({"fn_constructor_kwargs": stage.fn_constructor_kwargs})
+            kwargs["fn_constructor_kwargs"].update(
+                dict(
+                    input_column=self.input_column,
+                    output_column=self.output_column,
+                    carry_over=self.config.carry_over,
+                )
+            )
+
+            # Apply the stage.
+            dataset = dataset.map_batches(stage.fn, **kwargs)
+
+            # Rename the output column to the input column to chain the stages.
+            dataset = dataset.rename_columns({self.output_column: self.input_column})
+
+        if self.postprocess is not None:
+            dataset = dataset.map(self.postprocess)
+        return dataset
+
+    def append_stage(self, stage: StatefulStage):
+        """Append a stage before postprocess. The stage class name will be used as
+        the stage name. If there are multiple stages with the same type, a suffix
+        will be added to the stage name to avoid conflicts.
+
+        Args:
+            stage: The stage to append.
+        """
+        stage_name = type(stage).__name__
+
+        # When a processor has multiple stages with the same type,
+        # append a index suffix to the stage name to avoid conflicts.
+        if stage_name in self.stages:
+            num_same_type_stage = len([s for s in self.stages.values() if s is stage])
+            stage_name = f"{stage_name}_{num_same_type_stage}"
+        self.stages[stage_name] = stage
+
+    def list_stage_names(self) -> List[str]:
+        """List the stage names of this processor in order. Preprocess and postprocess
+        are not included.
+
+        Returns:
+            A list of stage names.
+        """
+        return list(self.stages.keys())
+
+    def get_stage_by_name(self, name: str) -> StatefulStage:
+        """Get a particular stage by its name. If the stage is not found,
+        a ValueError will be raised.
+
+        Args:
+            name: The stage name.
+
+        Returns:
+            The pipeline stage.
+        """
+        if name in self.stages:
+            return self.stages[name]
+        raise ValueError(f"Stage {name} not found")
+
+
+@DeveloperAPI
+class ProcessorBuilder:
+    """Build a processor based on the configuration."""
+
+    _registry: Dict[Type[ProcessorConfig], Callable] = {}
```

Comment:
is Type[ProcessorConfig] fully hashable (even if it's inheritance pattern)? or do you have to use class.__name__ or class.__qualname__ or sth similar?

Reply:
I see you are using `__name__` then the type here should be `Dict[str, Callable]`

Reply:
Ah good catch.